azure-kusto-data/azure/kusto/data/_token_providers.py

```diff
@@ -232,7 +233,7 @@ def _context_impl(self) -> dict:
     def _init_impl(self):
         pass
 
-    def _get_token_impl(self) -> dict:
+    def _get_token_impl(self) -> None:

```

Comment:
Instead if this make the class abstract and don't implement the abstract API?

Reply:
The base class is abstract.
I think not implementing an abstract method is an anti-pattern.
Explicit is better than implicit

Reply:
In that case do not change the interface (or change everything to optional[dict]

Reply:
Is it really changing the interface?
The base class has a weaker type, and this change makes it into a stronger type.
It still fits in a "Optional[dict]".
There are impls that only return dict, I don't think changing them to Optional[dict] will help

Reply:
Everything now returns Optional[Dict]